File: CSMApps/Vertebrae/v1.0.0/src/main/java/com/carrpod/vertebrae/network/KilosWebSocketClient.java
package com.carrpod.vertebrae.network;

import android.util.Log;

import com.carrpod.vertebrae.model.KilosSession;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.WebSocket;
import java.util.concurrent.CompletionStage;
import java.util.concurrent.TimeUnit;

public class KilosWebSocketClient {

    private static final String TAG = "KilosWebSocketClient";

    public interface WebSocketListener {
        void onOpen(String sessionId);
        void onMessage(String sessionId, String text);
        void onBinaryMessage(String sessionId, byte[] bytes);
        void onClosed(String sessionId, int code, String reason);
        void onError(String sessionId, String error);
    }

    private final String sessionId;
    private final WebSocketListener listener;
    private final HttpClient httpClient;
    private WebSocket webSocket;
    private boolean isConnecting = false;

    public KilosWebSocketClient(String sessionId, WebSocketListener listener) {
        this.sessionId = sessionId;
        this.listener = listener;
        this.httpClient = HttpClient.newBuilder()
                .build();
    }

    public void connect() {
        if (isConnecting || webSocket != null) return;
        isConnecting = true;

        String wsUrl = "wss://app.kilo.ai/cloud/chat?sessionId=" + sessionId;
        
httpClient.newWebSocketBuilder()
                .buildAsync(URI.create(wsUrl), new WebSocket.Listener() {
                    public void onOpen(WebSocket webSocket) {
                        KilosWebSocketClient.this.webSocket = webSocket;
                        isConnecting = false;
                        Log.d(TAG, "Connected: " + sessionId);
                        listener.onOpen(sessionId);
                        webSocket.request(1);
                    }

                    public CompletionStage<?> onText(WebSocket webSocket, CharSequence data, boolean last) {
                        listener.onMessage(sessionId, data.toString());
                        webSocket.request(1);
                        return null;
                    }

                    public CompletionStage<?> onBinary(WebSocket webSocket, byte[] data, boolean last) {
                        listener.onBinaryMessage(sessionId, data);
                        webSocket.request(1);
                        return null;
                    }

                    public void onError(WebSocket webSocket, Throwable error) {
                        isConnecting = false;
                        Log.e(TAG, "Error: " + sessionId, error);
                        listener.onError(sessionId, error.getMessage() != null ? error.getMessage() : "Connection failed");
                    }

                    public CompletionStage<?> onClose(WebSocket webSocket, int statusCode, String reason) {
                        KilosWebSocketClient.this.webSocket = null;
                        listener.onClosed(sessionId, statusCode, reason);
                        return null;
                    }
                })
                .exceptionally(throwable -> {
                    isConnecting = false;
                    Log.e(TAG, "Connection failed: " + sessionId, throwable);
                    listener.onError(sessionId, throwable.getMessage() != null ? throwable.getMessage() : "Connection failed");
                    return null;
                });
    }

    public void send(String text) {
        if (webSocket != null) {
            webSocket.sendText(text, true);
        } else {
            listener.onError(sessionId, "Not connected");
        }
    }

    public void send(byte[] bytes) {
        if (webSocket != null) {
            webSocket.sendBinary(java.nio.ByteBuffer.wrap(bytes), true);
        } else {
            listener.onError(sessionId, "Not connected");
        }
    }

    public void sendHeartbeat() {
        send("{\"type\":\"heartbeat\",\"sessionId\":\"" + sessionId + "\",\"t\":" + System.currentTimeMillis() + "}");
    }

    public void sendBrowserHeartbeat() {
        send("{\"type\":\"ping\",\"data\":{\"t\":" + System.currentTimeMillis() + "}}");
    }

    public void close() {
        if (webSocket != null) {
            webSocket.sendClose(WebSocket.NORMAL_CLOSURE, "Client disconnect").join();
            webSocket = null;
        }
        isConnecting = false;
        listener.onClosed(sessionId, 1000, "Client disconnect");
    }

    public boolean isConnected() {
        return webSocket != null;
    }
}

File: CSMApps/Vertebrae/v1.0.0/src/main/kotlin_backup/KilosWebSocketClient.kt
package com.carrpod.vertebrae.network

import android.content.Context
import android.util.Log
import okhttp3.*
import okio.ByteString
import java.util.concurrent.TimeUnit

class KilosWebSocketClient private constructor(
    private val sessionId: String,
    private val context: Context,
    private val listener: Listener
) {

    interface Listener {
        fun onOpen(sessionId: String)
        fun onMessage(sessionId: String, text: String)
        fun onBinaryMessage(sessionId: String, bytes: ByteArray)
        fun onClosed(sessionId: String, code: Int, reason: String)
        fun onError(sessionId: String, error: String)
    }

    private var webSocket: WebSocket? = null
    private val client = OkHttpClient.Builder()
        .pingInterval(15, TimeUnit.SECONDS)
        .build()
    private val request = Request.Builder()
        .url("wss://app.kilo.ai/cloud/chat?sessionId=$sessionId")
        .addHeader("User-Agent", "Mozilla/5.0 (Linux; Android 13) Vertebrae/1.0")
        .addHeader("Origin", "https://app.kilo.ai")
        .addHeader("Sec-WebSocket-Protocol", "chat.kilo.ai")
        .build()

    fun connect() {
        webSocket = client.newWebSocket(request, object : WebSocketListener() {
            override fun onOpen(webSocket: WebSocket, response: Response) {
                Log.d("VertebraeWS", "Session $sessionId connected")
                listener.onOpen(sessionId)
            }

            override fun onMessage(webSocket: WebSocket, text: String) {
                listener.onMessage(sessionId, text)
            }

            override fun onMessage(webSocket: WebSocket, bytes: ByteString) {
                listener.onBinaryMessage(sessionId, bytes.toByteArray())
            }

            override fun onClosing(webSocket: WebSocket, code: Int, reason: String) {
                webSocket.close(1000, null)
                listener.onClosed(sessionId, code, reason)
            }

            override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
                Log.d("VertebraeWS", "Session $sessionId closed: $code - $reason")
                listener.onClosed(sessionId, code, reason)
            }

            override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
                Log.e("VertebraeWS", "Session $sessionId error: ${t.message}")
                listener.onError(sessionId, t.message ?: "Unknown error")
            }
        })
    }

    fun send(text: String) {
        webSocket?.send(text)
    }

    fun send(bytes: ByteArray) {
        webSocket?.send(ByteString.of(bytes))
    }

    fun close() {
        webSocket?.close(1000, "Client closing")
    }

    fun isConnected(): Boolean {
        return webSocket != null
    }

    companion object {
        private val instances = mutableMapOf<String, KilosWebSocketClient>()

        fun getOrCreate(sessionId: String, context: Context, listener: Listener): KilosWebSocketClient {
            return instances.getOrPut(sessionId) {
                KilosWebSocketClient(sessionId, context.applicationContext, listener)
            }
        }

        fun remove(sessionId: String) {
            instances.remove(sessionId)?.close()
        }

        fun closeAll() {
            instances.values.forEach { it.close() }
            instances.clear()
        }
    }
}